### ISSUE TITLE
Support alternate clock sources

### ISSUE THREAD

**Issue by mj1856** [moment/moment#2337]
It would be a nice feature if the source of current time was pluggable, rather than always obtained from `new Date()`.  There should be a simple function that can be overwritten to provide custom functionality.

For example:
- I might want to get the current time from a server, rather than trusting the client's clock.
- It would be helpful during unit tests to set up a fake clock with fixed values.

I envision an interface similar to:

```javascript
// default implementation
moment.now = function() { return moment(new Date()); }

// an alternate implementation, using a starting value from elsewhere
var now = moment("2015-01-01T00:00:00Z");
setInterval(function(){ now.add(1,'s'); }, 1000);
moment.now = function() { return now; }

// a fixed value implementation, for use in unit tests
moment.now = function() { return moment("2015-01-01T00:00:00Z"); }
```

This seems like it would be very easy to implement.  Thoughts?

**Comment by mj1856** [comment]
Or, a better way to implement the second one might be:

```javascript
var offset = moment("2015-01-01T00:00:00Z").valueOf() - Date.now();
moment.now = function() { return moment(Date.now() + offset); }
```

**Comment by fracz** [comment]
+1

**Comment by jokesterfr** [comment]
I assume it's very useful for testing. +1

**Comment by jisaacks** [comment]
+1

**Comment by ichernev** [comment]
The function should return UNIX time, not some fancy moment/Date object, to break possible circular dependencies.

**Comment by mj1856** [comment]
@ichernev - good point.

**Comment by slaughter550** [comment]
@mj1856 @ichernev @jisaacks, @atogata has developed a proposed solution here https://github.com/moment/moment/pull/2766. Would you mind taking a look at it? Thanks!

**Comment by mj1856** [comment]
Closed, completed in #2766
